oak-run/src/main/java/org/apache/jackrabbit/oak/run/Downloader.java

```diff
@@ -0,0 +1,142 @@
+ */
+public class Downloader implements Closeable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Downloader.class);
+
+    private final ExecutorService executorService;
+
+    public Downloader(int concurrency) {
+        LOG.info("Initializing Downloader with max number of concurrent requests={}", concurrency);
+        this.executorService = new ThreadPoolExecutor(
+                (int) Math.ceil(concurrency * .1), concurrency, 60L, TimeUnit.SECONDS,
+                new LinkedBlockingQueue<>(),
+                new ThreadFactoryBuilder()
+                        .setNameFormat("downloader-%d")
+                        .setDaemon(true)
+                        .build()
+        );
+    }
+
+    public List<ItemResponse> download(List<Item> items) {
+        LOG.debug("Preparing to download {} items.\n{}", items.size(), items);
+        try {
+            return executorService
+                    .invokeAll(items.stream().map(DownloadWorker::new).collect(Collectors.toList()))
+                    .stream()
+                    .map(itemResponseFuture -> {
+                        try {
+                            return itemResponseFuture.get();
+                        } catch (InterruptedException | ExecutionException e) {
+                            throw new RuntimeException(e);
+                        }
+                    })
+                    .collect(Collectors.toList());
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        executorService.shutdown();
+    }
+
+    public static class Item {
+        public String source;
+        public String destination;
+
+        @Override
+        public String toString() {
+            return "Item{" +
+                    "source='" + source + '\'' +
+                    ", destination='" + destination + '\'' +
+                    '}';
+        }
+    }
+
+    public static class ItemResponse {
+        public final Item item;
+        public boolean failed;
+        public long size;
+        public long time;
+        public Throwable throwable;
+
+        public ItemResponse(Item item) {
+            this.item = item;
+        }
+    }
+
+    private static class DownloadWorker implements Callable<ItemResponse> {
+
+        private final Item item;
+
+        DownloadWorker(Item item) {
+            this.item = item;
+        }
+
+        @Override
+        public ItemResponse call() {
+            ItemResponse response  = new ItemResponse(item);
+            long t0 = System.currentTimeMillis();
+            try {
+                URL sourceUrl = new URL(item.source);
+                File destinationFile = new File(item.destination);
+                destinationFile.getParentFile().mkdirs();
+                try (ReadableByteChannel byteChannel = Channels.newChannel(sourceUrl.openStream());
```

Comment:
What is the connect and read timeout? Do we want to configure those explicitly? For inter region transfers from the Azure blobstore, the response should be always under a few hundred milliseconds. So probably the timeouts should be relatively short. I'm not sure what are the default timeouts, but I have the idea that they are quite high. I saw somewhere 15s for connect, 60s for read. This may be excessive in this case.

Reply:
By default, it seems that timeouts in `HttpURLConnection` are zero (no timeouts).
I have introduced 60 seconds for connect and 60 minutes for read.

We cannot make assumptions on the Azure blobstore for various reasons. In the past, we have seen huge differences between executions, and depending on the number of requests, Azure can decide to start throttling requests.

Reply:
The 60 minutes for read seems excessive. Can there be situations where a connection to the blob store stops sending data for, let's say, 50 minutes and then it restarts? That is, if a connection is not sending data for more than a few minutes, is it reasonable to expect that it can resume? 

I'm not familiar with Azure blob store, I'm writing out of experience with S3, where in my previous job we had trouble when we used very large timeouts, especially the read timeout. The system seemed to hang when it hit these timeouts and I never seen any situation where a large read timeout was useful. Usually if the data flow stoped for more than a few tens of seconds, it would not recover. In the end, we were setting a generous connect timeout (but not more than a couple of minutes) and a read timeout on the order of tens of seconds, with the assumption that once a connection is established, the data will flow at a steady pace.

Reply:
I see your point. I made both `connect` and `read` timeouts configurable from the command options.